File: backend/src/crypto/providers/ciphers.py
"""Grade A: AES-256-GCM AEAD for message and file content.

A fresh, cryptographically random 96-bit nonce is generated for every
encrypt call (Constitution: Secure Random / no reused nonces).
"""

import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from src.crypto.interfaces import FileCipher, MessageCipher

_NONCE_LENGTH_BYTES = 12


def _encrypt(key: bytes, plaintext: bytes, associated_data: bytes) -> tuple[bytes, bytes]:
    nonce = os.urandom(_NONCE_LENGTH_BYTES)
    ciphertext = AESGCM(key).encrypt(nonce, plaintext, associated_data)
    return ciphertext, nonce


def _decrypt(key: bytes, ciphertext: bytes, nonce: bytes, associated_data: bytes) -> bytes:
    return AESGCM(key).decrypt(nonce, ciphertext, associated_data)


class AesGcmMessageCipher(MessageCipher):
    def encrypt(
        self, key: bytes, plaintext: bytes, *, associated_data: bytes = b""
    ) -> tuple[bytes, bytes]:
        return _encrypt(key, plaintext, associated_data)

    def decrypt(
        self, key: bytes, ciphertext: bytes, nonce: bytes, *, associated_data: bytes = b""
    ) -> bytes:
        return _decrypt(key, ciphertext, nonce, associated_data)


class AesGcmFileCipher(FileCipher):
    def encrypt(
        self, key: bytes, plaintext: bytes, *, associated_data: bytes = b""
    ) -> tuple[bytes, bytes]:
        return _encrypt(key, plaintext, associated_data)

    def decrypt(
        self, key: bytes, ciphertext: bytes, nonce: bytes, *, associated_data: bytes = b""
    ) -> bytes:
        return _decrypt(key, ciphertext, nonce, associated_data)

File: frontend/src/crypto/providers/ciphers.ts
/**
 * Grade A: AES-256-GCM AEAD for message and file content, via
 * `@noble/ciphers` (kept alongside the ML-KEM/ML-DSA providers for a
 * consistent, audited, cross-browser implementation rather than relying on
 * SubtleCrypto directly). A fresh random 96-bit nonce is generated per
 * encrypt call.
 */

import { gcm } from "@noble/ciphers/aes";
import type { EncryptedPayload, FileCipher, MessageCipher } from "../interfaces";

const NONCE_LENGTH_BYTES = 12;

function encryptWith(
  key: Uint8Array,
  plaintext: Uint8Array,
  associatedData: Uint8Array,
): EncryptedPayload {
  const nonce = crypto.getRandomValues(new Uint8Array(NONCE_LENGTH_BYTES));
  const ciphertext = gcm(key, nonce, associatedData).encrypt(plaintext);
  return { ciphertext, nonce };
}

function decryptWith(
  key: Uint8Array,
  payload: EncryptedPayload,
  associatedData: Uint8Array,
): Uint8Array {
  return gcm(key, payload.nonce, associatedData).decrypt(payload.ciphertext);
}

export const aes256GcmMessageCipher: MessageCipher = {
  async encrypt(key, plaintext, associatedData = new Uint8Array()) {
    return encryptWith(key, plaintext, associatedData);
  },
  async decrypt(key, payload, associatedData = new Uint8Array()) {
    return decryptWith(key, payload, associatedData);
  },
};

export const aes256GcmFileCipher: FileCipher = {
  async encrypt(key, plaintext, associatedData = new Uint8Array()) {
    return encryptWith(key, plaintext, associatedData);
  },
  async decrypt(key, payload, associatedData = new Uint8Array()) {
    return decryptWith(key, payload, associatedData);
  },
};
